pkg/assetfetcher/http.go

```diff
@@ -0,0 +1,162 @@
+	errChan := make(chan error, 1)
+	defer close(errChan)
+
+	// have to page through to get all the assets, start with page 0
+	// make the first call to Nexpose to get the total number of pages we'll
+	var currentPage = 0
+	req, err := newNexposeSiteAssetsRequest(c.Host, siteID, currentPage, c.PageSize)
+	if err != nil {
+		errChan <- err
+		return nil, errChan
+	}
+
+	res, err := c.HTTPClient.Do(req.WithContext(ctx))
+	if err != nil {
+		errChan <- &NexposeHTTPRequestError{err, req.URL.String()}
+		return nil, errChan
+	}
+	defer res.Body.Close()
+	respBody, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		errChan <- &ErrorReadingNexposeResponse{err, req.URL.String()}
+		return nil, errChan
+	}
+
+	var siteAssetResp SiteAssetsResponse
+	if err := json.Unmarshal(respBody, &siteAssetResp); err != nil {
+		errChan <- &ErrorParsingJSONResponse{err, req.URL.String()}
+		return nil, errChan
+	}
+
+	pagedAssetChan := make(chan domain.Asset, siteAssetResp.Page.TotalResources)
+	pagedErrChan := make(chan error, siteAssetResp.Page.TotalResources)
+
+	for _, asset := range siteAssetResp.Resources {
+		pagedAssetChan <- asset
+	}
+
+	// We've gotten the first page (page 0) and added the assets to the channel,
+	// so here we'll get the TotalPages of assets that Nexpose has, then paginate
+	// through from page 1 to TotalPages to get all the assets
+	totalPages := siteAssetResp.Page.TotalPages
+	var wg sync.WaitGroup
+	for currentPage := 1; currentPage < totalPages; currentPage++ {
+		wg.Add(1)
+		go c.makeRequest(ctx, &wg, siteID, currentPage, pagedAssetChan, pagedErrChan)
+	}
+
+	go func() {
+		defer close(pagedErrChan)
+		defer close(pagedAssetChan)
+		wg.Wait()
+	}()
+
+	return pagedAssetChan, pagedErrChan
+}
+
+func (c *NexposeAssetFetcher) makeRequest(ctx context.Context, wg *sync.WaitGroup, siteID string, page int, assetChan chan domain.Asset, errChan chan error) {
+	defer wg.Done()
+	req, err := newNexposeSiteAssetsRequest(c.Host, siteID, page, c.PageSize)
+	if err != nil {
+		errChan <- err
+		return
+	}
+	res, err := c.HTTPClient.Do(req.WithContext(ctx))
+	if err != nil {
+		errChan <- &NexposeHTTPRequestError{err, req.URL.String()}
+		return
+	}
+	defer res.Body.Close()
+	respBody, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		errChan <- &ErrorReadingNexposeResponse{err, req.URL.String()}
+		return
+	}
+	var siteAssetResp SiteAssetsResponse
+	if err := json.Unmarshal(respBody, &siteAssetResp); err != nil {
+		errChan <- &ErrorParsingJSONResponse{err, req.URL.String()}
+		return
+	}
+	for _, asset := range siteAssetResp.Resources {
+		assetChan <- asset
+	}
+}
+
+// newNexposeSiteAssetsRequest builds URL we'll use to make the request to Nexpose's site assets endpoint
+func newNexposeSiteAssetsRequest(host string, siteID string, page int, size int) (*http.Request, error) {
+	u, err := url.Parse(host)
+	if err != nil {
+		return nil, &URLParsingError{err, host}
```

Comment:
Why use a custom error here? While our team has used error types to inform upstream behavior, in general it's not desired to use typed errors. Each time you do so you are adding another element to a function's public interface. If the goal is to simply add addition context to an error, you may want to look into something like `errors.Wrap`. see: https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully

Reply:
We can talk about this more IRL. I'm not sure our team agrees on this and I'd like to save improvements for a different issue. Thanks for pointing it out